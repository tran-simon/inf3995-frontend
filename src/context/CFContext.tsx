import React, {
  createContext,
  ReactNode,
  useCallback,
  useEffect,
  useState,
} from 'react';
import Crazyflie from '../model/Crazyflie';
import { noop } from 'lodash';
import { SetState } from '../utils/utils';

const BACKEND_URL =
  process.env.REACT_APP_BACKEND_URL ?? 'http://localhost:5000';

interface ICFContext {
  cfList: Crazyflie[];

  scan: () => Promise<Response | void>;
  updateStats: () => Promise<Response | void>;

  backendUrl?: string;
  setBackendUrl: SetState<string | undefined>;

  backendDisconnected?: boolean;

  refreshRate?: number;
  setRefreshRate: SetState<number>;

  land: () => Promise<Response | void>;
  takeoff: () => Promise<Response | void>;
}

const DefaultCfContext: ICFContext = {
  cfList: [],
  scan: async () => {},
  updateStats: async () => {},
  backendUrl: BACKEND_URL,
  setBackendUrl: noop,

  setRefreshRate: noop,

  land: async () => {},
  takeoff: async () => {},
};

const CFContext = createContext<ICFContext>(DefaultCfContext);

export const CFProvider = ({
  children,
  ...props
}: Partial<ICFContext> & {
  children: ReactNode;
}) => {
  const [refreshRate, setRefreshRate] = useState<number>(
    props.refreshRate ?? 0,
  );
  const [backendUrl, setBackendUrl] = useState<string | undefined>(
    props.backendUrl ?? BACKEND_URL,
  );
  const [backendDisconnected, setBackendDisconnected] = useState(
    props.backendDisconnected ?? !!backendUrl,
  );
  const [cfList, setCfList] = useState<Crazyflie[]>(props.cfList || []);

  const scan = useCallback(async () => {
    if (!backendDisconnected) {
      return fetch(`${backendUrl}/scan`).then((res) => {
        if (res.ok) {
          res.json().then((value) => {
            if (Array.isArray(value)) {
              setCfList(value);
            }
          });
        }
      });
    }
  }, [backendUrl, backendDisconnected]);

  const updateStats = useCallback(async () => {
    if (!backendDisconnected) {
      return fetch(`${backendUrl}/updateStats`).then((res) => {
        if (res.ok) {
          return res.json().then((val) => {
            if (Array.isArray(val)) {
              setCfList(val);
            }
          });
        }
      });
    }
  }, [backendUrl, backendDisconnected]);

  useEffect(() => {
    fetch(`${backendUrl}/liveCheck`)
      .then((response) => {
        setBackendDisconnected(!response.ok);
      })
      .then(updateStats)
      .catch(() => {
        setBackendDisconnected(true);
      });
  }, [backendUrl, updateStats]);

  useEffect(() => {
    if (!backendDisconnected && !!refreshRate) {
      const refresh = setInterval(() => {
        updateStats();
      }, refreshRate);
      return () => clearInterval(refresh);
    }
  }, [refreshRate, backendUrl, backendDisconnected, updateStats]);

  const takeoff = async () => {
    if (!backendDisconnected) {
      return fetch(`${backendUrl}/takeOff`);
    }
  };
  const land = async () => {
    if (!backendDisconnected) {
      return fetch(`${backendUrl}/land`);
    }
  };

  return (
    <CFContext.Provider
      value={{
        ...DefaultCfContext,
        ...props,
        backendUrl,
        setBackendUrl,
        backendDisconnected,
        scan,
        updateStats,
        refreshRate,
        setRefreshRate,
        cfList,
        takeoff,
        land,
      }}
    >
      {children}
    </CFContext.Provider>
  );
};

export default CFContext;
