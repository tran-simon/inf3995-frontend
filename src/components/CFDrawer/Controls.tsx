import React, { ReactNode, useContext, useState } from 'react';
import {
  CircularProgress,
  Divider,
  List,
  ListItem,
  ListItemIcon,
  ListItemText,
  TextField,
} from '@material-ui/core';
import RefreshIcon from '@material-ui/icons/Refresh';
import CFContext from '../../context/CFContext';
import ConnectedIcon from '@material-ui/icons/WifiTethering';
import DiconnectedIcon from '@material-ui/icons/PortableWifiOff';
import ToggleOffOutlinedIcon from '@material-ui/icons/ToggleOffOutlined';
import ToggleOnOutlinedIcon from '@material-ui/icons/ToggleOnOutlined';

type ControlSectionProps = {
  icon?: ReactNode;
  onClick?: () => Promise<any>;
  children?: ReactNode;
};
const ControlSection = ({
  icon,
  onClick = async () => {},
  children,
}: ControlSectionProps) => {
  const [loading, setLoading] = useState(false);

  return (
    <ListItem
      button
      onClick={() => {
        setLoading(true);
        onClick().finally(() => setLoading(false));
      }}
    >
      <ListItemIcon>
        {!loading ? icon : <CircularProgress size="1rem" />}
      </ListItemIcon>
      <ListItemText>{children}</ListItemText>
    </ListItem>
  );
};

export const Controls = () => {
  const {
    scan,
    backendDisconnected,
    backendUrl,
    setBackendUrl,
    setRefreshRate,
    refreshRate,
    simulation,
    reset,
  } = useContext(CFContext);

  return (
    <List>
      <ControlSection icon={<RefreshIcon />} onClick={scan}>
        Balayer
      </ControlSection>

      <ControlSection
        onClick={async () => {
          setRefreshRate(refreshRate ? 0 : 1000);
        }}
        icon={
          refreshRate ? <ToggleOnOutlinedIcon /> : <ToggleOffOutlinedIcon />
        }
      >
        Rafra√Æchissement Automatique
      </ControlSection>

      <ControlSection
        icon={simulation ? <ToggleOnOutlinedIcon /> : <ToggleOffOutlinedIcon />}
        onClick={() => reset(!simulation)}
      >
        Mode Simulation
      </ControlSection>

      <Divider />

      <ListItem>
        <ListItemIcon>
          {backendDisconnected ? <DiconnectedIcon /> : <ConnectedIcon />}
        </ListItemIcon>

        <TextField
          label={'URL du serveur'}
          value={backendUrl}
          onChange={(event) => {
            setBackendUrl(event.target.value);
          }}
          error={backendDisconnected}
        />
      </ListItem>
    </List>
  );
};
