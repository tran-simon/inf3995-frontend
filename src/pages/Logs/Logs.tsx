import React, {
  useCallback,
  useContext,
  useEffect,
  useRef,
  useState,
} from 'react';
import {
  Button,
  Dialog,
  DialogActions,
  DialogContent,
  DialogContentText,
  DialogProps,
  DialogTitle,
  Typography,
} from '@material-ui/core';
import CFContext from '../../context/CFContext';
import RefreshIcon from '@material-ui/icons/Refresh';

const Logs = (props: DialogProps) => {
  const { setShowLogs, refreshLogs } = useContext(CFContext);
  const [logs, setLogs] = useState('');
  const endMessageRef = useRef<any>(null);

  const refresh = useCallback(() => {
    refreshLogs().then((res) => {
      if (typeof res === 'string') {
        setLogs(res);
      }
    });
  }, [refreshLogs, setLogs]);

  useEffect(() => {
    if (props.open) {
      refresh();
    }
  }, [refresh, props.open]);

  useEffect(() => {
    if (endMessageRef.current != null) {
      endMessageRef.current.scrollIntoView({});
    }
  }, []);

  return (
    <Dialog
      onClose={() => setShowLogs(false)}
      fullWidth
      maxWidth="md"
      scroll={'paper'}
      {...props}
    >
      <DialogTitle>Journal des messages</DialogTitle>
      <DialogContent dividers>
        <DialogContentText style={{ whiteSpace: 'pre-line' }}>
          {logs}
          <Typography
            component="span"
            ref={endMessageRef}
            color="textPrimary"
            style={{ textAlign: 'center', margin: '1rem' }}
          >
            -- Fin des messages --
          </Typography>
        </DialogContentText>
      </DialogContent>
      <DialogActions>
        <Button
          onClick={() => {
            refresh();
          }}
        >
          <RefreshIcon />
          Rafra√Æchir
        </Button>
      </DialogActions>
    </Dialog>
  );
};
export default Logs;
