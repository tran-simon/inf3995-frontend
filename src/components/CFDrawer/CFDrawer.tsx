import React, { ReactElement, useContext, useState } from 'react';
import {
  Box,
  Button,
  Dialog,
  DialogActions,
  DialogContent,
  DialogContentText,
  DialogTitle,
  Divider,
  Drawer,
  DrawerProps,
  Grid,
  List,
  ListItem,
  ListItemIcon,
  ListItemText,
  TextField,
  Tooltip,
  Typography,
  useTheme,
} from '@material-ui/core';
import CFContext from '../../context/CFContext';
import { buildStyles, CircularProgressbar } from 'react-circular-progressbar';
import 'react-circular-progressbar/dist/styles.css';
import StandbyIcon from '@material-ui/icons/HourglassEmpty';
import InMissionIcon from '@material-ui/icons/AirplanemodeActive';
import CrashedIcon from '@material-ui/icons/AirplanemodeInactive';
import Crazyflie, { State } from '../../model/Crazyflie';
import { Controls } from './Controls';

const StateIcon = ({ crazyflie }: { crazyflie: Crazyflie }) => {
  let stateIcon: ReactElement | null;
  let stateText = '';
  switch (crazyflie.state) {
    case State.inMission:
      stateText = 'En Mission';
      stateIcon = <InMissionIcon />;
      break;
    case State.crashed:
      stateText = 'Écrasé';
      stateIcon = <CrashedIcon />;
      break;
    default:
      stateText = 'En Attente';
      stateIcon = <StandbyIcon />;
      break;
  }

  return stateIcon && <Tooltip title={stateText}>{stateIcon}</Tooltip>;
};

export const CFDrawer = ({ children, ...props }: DrawerProps) => {
  const { cfList } = useContext(CFContext);
  const [openCfId, setOpenCfId] = useState<string | null>(null);
  const { palette } = useTheme();

  return (
    <Drawer variant="permanent" anchor="left" {...props}>
      {children}
      <Box flexGrow={1}>
        <List>
          {Object.keys(cfList).map((droneId, i) => {
            const crazyflie = cfList[droneId];
            if (!crazyflie) {
              return null;
            }
            const batteryPer = crazyflie.battery ?? 0;

            return (
              <Tooltip
                key={i}
                title={
                  !crazyflie.initialPosition
                    ? `Erreur: La position initiale n\'a pas été spécifiée pour ${droneId}`
                    : droneId
                }
              >
                <ListItem button onClick={() => setOpenCfId(droneId)}>
                  <ListItemIcon
                    style={{
                      width: '40px',
                      minWidth: 'unset',
                      marginRight: '1rem',
                    }}
                  >
                    <CircularProgressbar
                      value={batteryPer}
                      text={`${batteryPer.toFixed(2)} %`}
                      background
                      styles={
                        !crazyflie.initialPosition
                          ? buildStyles({
                              backgroundColor: palette.error.main,
                              textColor: palette.error.contrastText,
                              trailColor: palette.error.light,
                              pathColor: palette.error.dark,
                            })
                          : undefined
                      }
                    />
                  </ListItemIcon>
                  <ListItemText
                    secondary={`Vitesse: ${crazyflie.speed?.toFixed(2) || ''}`}
                    color={
                      !crazyflie.initialPosition
                        ? palette.error.main
                        : undefined
                    }
                  >
                    <Typography
                      color={!crazyflie.initialPosition ? 'error' : undefined}
                    >
                      {droneId}
                    </Typography>
                  </ListItemText>
                  <ListItemIcon style={{ justifyContent: 'flex-end' }}>
                    <StateIcon crazyflie={crazyflie} />
                  </ListItemIcon>
                </ListItem>
              </Tooltip>
            );
          })}
        </List>
      </Box>
      {openCfId && (
        <CFDialog
          droneId={openCfId}
          close={() => setOpenCfId(null)}
          save={(cf) => {}}
        />
      )}
      <Controls />
    </Drawer>
  );
};

const CFDialog = ({
  droneId,
  close,
}: {
  droneId: string;
  close: () => void;
  save: (crazyflie: Crazyflie) => void;
}) => {
  const { cfList, setCfList } = useContext(CFContext);
  const crazyflie = cfList[droneId];
  const [x, setX] = useState<string>(
    String(crazyflie?.initialPosition?.x ?? ''),
  );
  const [y, setY] = useState<string>(
    String(crazyflie?.initialPosition?.y ?? ''),
  );
  if (!crazyflie) {
    return null;
  }
  return (
    <Dialog open={!!crazyflie} onClose={() => close()}>
      <DialogTitle>Crazyflie: {droneId}</DialogTitle>
      <DialogContent>
        <List>
          <ListItem>
            <DialogContentText>
              Batterie: {crazyflie.battery?.toFixed(2)} %
            </DialogContentText>
          </ListItem>
          <ListItem>
            <DialogContentText>Vitesse: {crazyflie.speed}</DialogContentText>
          </ListItem>
          <ListItem>
            <DialogContentText style={{}}>
              État: <StateIcon crazyflie={crazyflie} />
            </DialogContentText>
          </ListItem>
          <Divider />
          <ListItem>
            <Grid container spacing={2}>
              <Grid item xs={12}>
                <Typography variant="h6">Position Initiale</Typography>
              </Grid>
              <Grid item xs={6}>
                <TextField
                  label="X"
                  value={x}
                  onChange={(e) => {
                    setX(e.target.value.replace(/[^0-9.,-]/, ''));
                  }}
                  onBlur={(e) => {
                    setX(String(+e.target.value));
                  }}
                  type="number"
                />
              </Grid>
              <Grid item xs={6}>
                <TextField
                  label="Y"
                  value={y}
                  onChange={(e) => {
                    setY(e.target.value.replace(/[^0-9.,-]/, ''));
                  }}
                  onBlur={(e) => {
                    setY(String(+e.target.value));
                  }}
                  type="number"
                />
              </Grid>
            </Grid>
          </ListItem>
        </List>
      </DialogContent>
      <DialogActions>
        <Button onClick={() => close()}>Annuler</Button>
        <Button
          disabled={x === '' || y === ''}
          onClick={() => {
            const xNum = +x,
              yNum = +y;
            crazyflie.initialPosition = {
              x: xNum,
              y: yNum,
            };
            setCfList({ ...cfList });
            close();
          }}
        >
          Confirmer
        </Button>
      </DialogActions>
    </Dialog>
  );
};
