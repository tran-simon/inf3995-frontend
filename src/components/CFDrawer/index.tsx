import { CFDrawer } from './CFDrawer';
export default CFDrawer;
