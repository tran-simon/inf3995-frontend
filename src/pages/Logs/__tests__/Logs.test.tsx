import React, { FunctionComponent, ReactElement } from 'react';
import CFContext, { CFProvider } from '../../../context/CFContext';
import { CFDrawer } from '../../../components/CFDrawer/CFDrawer';
import { render, waitFor } from '@testing-library/react';
import Logs from '../Logs';

describe('Logs', () => {
  let refreshLogs: () => Promise<string>,
    setShowLogs: () => void,
    LogsComp: FunctionComponent<any>;

  beforeEach(() => {
    refreshLogs = jest.fn(() => {
      return Promise.resolve('logs');
    });
    setShowLogs = jest.fn();

    LogsComp = ({ open = false }: { open?: boolean }) => (
      <CFContext.Provider
        value={
          {
            setShowLogs,
            refreshLogs,
          } as any
        }
      >
        <Logs open={open} />
      </CFContext.Provider>
    );
  });

  it('can match snapshot', async () => {
    await waitFor(() => {
      expect(render(<LogsComp />).asFragment()).toMatchSnapshot();
    });
  });

  it('refreshes on open', async () => {
    const { rerender } = render(<LogsComp />);
    expect(refreshLogs).not.toHaveBeenCalled();

    await waitFor(() => {
      rerender(<LogsComp open />);
      expect(refreshLogs).toHaveBeenCalled();
    });
  });
});
