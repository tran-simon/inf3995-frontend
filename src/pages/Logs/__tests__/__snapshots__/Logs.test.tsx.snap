// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Logs can match snapshot 1`] = `<DocumentFragment />`;
